src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -2,48 +2,31 @@
 
 import nextstep.qna.CannotDeleteException;
 import nextstep.qna.NotFoundException;
-import nextstep.qna.domain.*;
+import nextstep.qna.domain.AnswerRepository;
+import nextstep.qna.domain.DeleteHistory;
+import nextstep.qna.domain.Question;
+import nextstep.qna.domain.QuestionRepository;
 import nextstep.users.domain.NsUser;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import javax.annotation.Resource;
 import java.time.LocalDateTime;
-import java.util.ArrayList;
 import java.util.List;
 
 @Service("qnaService")
 public class QnAService {
     @Resource(name = "questionRepository")
     private QuestionRepository questionRepository;
 
-    @Resource(name = "answerRepository")
-    private AnswerRepository answerRepository;
-
     @Resource(name = "deleteHistoryService")
     private DeleteHistoryService deleteHistoryService;
 
     @Transactional
     public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
         Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
-
-        List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
+        question.delete(loginUser);
+        List<DeleteHistory> deleteHistories = DeleteHistory.of(question);
         deleteHistoryService.saveAll(deleteHistories);
```

Comment:
```suggestion
        List<DeleteHistory> deleteHistories = question.delete(loginUser);
        deleteHistoryService.saveAll(deleteHistories);
```

삭제 이력 저장을 깜빡하지 않도록 하기 위해,
`delete()` 메소드에서 삭제 이력을 반환하도록 설계해보는 건 어떨까요?

Reply:
약간 고민했던 부분인데요.
CQRS 관점으로 봤을 때 커맨드에서 리턴 값까지 마련하면 역할이 두 개가 되는 것이 아닐까 하는 생각이 들었습니다.
커맨드는 void를 반환해야한다는 절대적인 규칙은 아니지만, 순수성 관점으로 봤을 때는 Question(도메인)이 DeleteHistory(다른 도메인)를 알아야하는게 책임이 커져서 조금 이상하다고 생각했습니다.
삭제 후 처리 요구사항이 많아져서 DeleteHistory 와 같은 도메인이 많이 필요해지면 DeleteResult 클래스 같은 것을 만들어서 리턴하고 DeleteResult에 DeleteHistory, ... 이런 것들을 넣어서 반환하는 것으로 가게되는 것은 아닐까 등의 생각이 들어서 void가 차라리 지금은 적절하다는 생각을 했습니다.

물론 망상이며 아직 일어나지 않는 일에 대한 생각이긴 하고 조금 많이 나아가서 이벤트로 설계하면 도메인의 순수성을 잘 챙기며 삭제 이력 저장이라는 로직도 Question 도메인의 관심사 밖으로 나가기 때문에 좋아져 보이긴 했습니다.

다만, 과제의 목적이 레거시 코드 리팩토링에 있으니 거기까지는 오버같아 서비스에서 마무리 짓는 방향으로 하게 되었습니다.
그럼에도 불구하고 삭제이력을 반환하는게 더 나은 설계로 보신다면 리뷰 후에 진행하겠습니다~

Reply:
네 동교님 말씀도 맞습니다!
말씀하신 것처럼 도메인 간 결합 이슈는 분명히 존재하기 때문에, 저 역시 이벤트 기반의 접근을 더 선호하긴 합니다.

항상 정답이 있는 건 아니고, 결국엔 상황이나 규모에 따라 트레이드오프가 있는 것 같아요.
특히 소규모 환경에서는 도메인 간 참조를 어느 정도 타협하고 허용하는 것도 하나의 방법이라고 생각합니다.

아마 이벤트 관련 경험도 있으실 것 같은데요
예전에 간단한 도메인 이벤트를 적용했을 때, 협업 중 소통이 부족해서 이벤트가 Fan-Out 구조로 확장되지 못하고 1:1로만 처리되는 상황도 있었고,
이벤트를 구독하는 핸들러가 여기저기 흩어져 있어서, 실제로는 단순한 로직인데도 이후에 어떤 처리가 일어나는지 직관적으로 파악하기 어려웠던 점도 있었습니다.

이번 미션의 핵심은 '복잡한 서비스 로직을 도메인 모델로 옮겨보자'는 부분이라
가능하다면 도메인 간 협력만으로도 해결할 수 있다는 방향을 공유드리고 싶었습니다.

리뷰 내용은 꼭 반영하실 필요는 없이 참고만 해주셔도 됩니다!
좋은 의견 감사드려요! 👍 😄